### PR TITLE
HW2

### PR DESCRIPTION
**Pull request by IgorRyabtsev**

### CODE REVIEW

**Comment by senia-psm** [comment]
Why is this `PR` closed?

File: src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java [hunk 1]
```diff
@@ -24,7 +24,7 @@ public void personHasNotEmptyLastNameAndFirstName0() {
     private Predicate<Person> negate1(Predicate<Person> test) {
         return p -> {
             // TODO
-            throw new UnsupportedOperationException();
+            return test.test(p);
```
Comment (senia-psm) [inline src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java]: Incorrect.

File: src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java [hunk 2]
```diff
@@ -33,7 +33,7 @@ public void personHasNotEmptyLastNameAndFirstName0() {
     private Predicate<Person> validateFirstNameAndLastName(Predicate<Person> t1, Predicate<Person> t2) {
         return p -> {
             // TODO
-            throw new UnsupportedOperationException();
+            return t1.test(p) ? false : !t2.test(p);
```
Comment (senia-psm) [inline src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java]: Incorrect

File: src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java [hunk 3]
```diff
@@ -56,25 +56,25 @@ public void personHasNotEmptyLastNameAndFirstName1() {
     // negate: (T -> boolean) -> (T -> boolean)
     private <T> Predicate<T> negate(Predicate<T> test) {
         // TODO
-        throw new UnsupportedOperationException();
+        return p -> !test.test(p);
```
Comment (senia-psm) [inline src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java]: Correct. `negate1` should be the same.

File: src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java [hunk 4]
```diff
@@ -56,25 +56,25 @@ public void personHasNotEmptyLastNameAndFirstName1() {
...
-        throw new UnsupportedOperationException();
+        return p -> !test.test(p);
     }
 
     // TODO
     // and: (T -> boolean, T -> boolean) -> (T -> boolean)
     private <T> Predicate<T> and(Predicate<T> t1, Predicate<T> t2) {
         // TODO
-        throw new UnsupportedOperationException();
+        return p -> t1.test(p) && t2.test(p);
```
Comment (senia-psm) [inline src/test/java/lambda/part2/exercise/FunctionCombinationExercise.java]: Correct. `validateFirstNameAndLastName` should be the same.

**Comment by IgorRyabtsev** [comment]
I've mixed hw1 and hw2.

**Pull request closed without merging**